File: rust/querier/dialogue_tree.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, PartialEq)]
/// Talk struct represents somone talking in a dialogue.
pub struct Talk<'a> {
    /// Who field represents who's talking.
    pub who: &'a str,
    /// What field represents what's being said.
    pub what: &'a str,
}

#[derive(Debug, Serialize, Deserialize, PartialEq)]
/// Choice struct represents the choice in a dialogue.
pub struct Choice<'a> {
    #[serde(borrow)]
    /// What field represents what's being said.
    pub what: &'a str,
    /// Next field represents the id of the next Node in the database.
    pub next: i32,
}

#[derive(Debug, Serialize, Deserialize, PartialEq)]
/// StoryNode is a struct to represent the json format of a node data field.
pub struct StoryNode<'a> {
    #[serde(borrow)]
    /// The dialogue for the story node.
    pub story: Vec<Talk<'a>>,
    #[serde(borrow)]
    /// The choices for the next node if it exists.
    pub select: Option<Vec<Choice<'a>>>,
    /// Has this story node been visited before.
    pub visited: bool,
}

File: engine/models/dialogue_tree.cc
#include "dialogue_tree.hh"

void to_json(json& j, const Talk& talk) {
  j = json{
      {"who", talk.who},
      {"what", talk.what},
  };
}

void from_json(const json& j, Talk& talk) {
  j.at("who").get_to(talk.who);
  j.at("what").get_to(talk.what);
}

void to_json(json& j, const Choice& choice) {
  j = json{
      {"what", choice.what},
      {"next", choice.next},
  };
}

void from_json(const json& j, Choice& choice) {
  j.at("what").get_to(choice.what);
  j.at("next").get_to(choice.next);
}

void to_json(json& j, const Story& story) {
  j = json{
      {"dialogue", story.dialogue},
      {"choices", story.choices},
      {"visited", story.visited},
  };
}

void from_json(const json& j, Story& story) {
  j.at("dialogue").get_to(story.dialogue);
  j.at("choices").get_to(story.choices);
  j.at("visited").get_to(story.visited);
}